climada/hazard/base.py

```diff
@@ -49,6 +49,23 @@ class Hazard(HazardIO, HazardPlot):
     Contains events of some hazard type defined at centroids. Loads from
     files with format defined in FILE_EXT.
 
+    Attention
+    ---------
+    This class uses instances of
+    `scipy.sparse.csr_matrix
+    <https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html>`_
+    to store :py:attr:`intensity` and :py:attr:`fraction`. This data types comes with
+    its particular pitfalls. Depending on how the objects are instantiated and modified,
+    a matrix might end up in a "non-canonical" state. In this state, its ``.data``
+    attribute does not necessarily represent the values apparent in the final matrix.
+    In particular, a "non-canonical" matrix may store "duplicates", i.e. multiple values
+    that map to the same matrix position. This is supported, and the default behavior is
+    to sum up these values. To avoid any inconsistencies, call :py:meth:`check_matrices`
+    once you inserted all data. This will explicitly sum all values at the same matrix
```

Comment:
Do we want to recommend checking that if the class does check it anyway? I am a bit sceptical with these `check_object` methods that we have in climada. Imho these should be handled in the `init` and that is it.

Reply:
hmmm... not yet fully convinced. If the user messes up an object, which the user always can in Python, then it is their problem. I would rather have fewer checks at random places in the code that hide bad habits and save people from doing things they should not do, and instead force good habits by having checks in the inits.

Reply:
I generally agree that we should only check if necessary. However, the MDR calculation operates on the `data` attribute. It is therefore essential that canonical format is ensured before starting the impact calculation.

If you really want to boil it down to a minimum, I would suggest to remove the pruning from the Hazard setters and instead only call `check_matrices` from `ImpactCalc`. But this might increase confusion again, because consistency is never ensured *before* starting `ImpactCalc`.

Another thing to keep in mind is that `sum_duplicates` is O(N^2) for a non-canonical matrix, and a no-op O(1) for a canonical matrix, and `eliminate_zeros` is O(N), where N is the number of stored non-zeros entries.

Reply:
Good points, and thanks for looking up the time cost of the methods.

Ok, let's try it like this. Maybe we can keep an eye on this change in the separate efforts to optimize computation times for ImpactCalc.

Reply:
In this case, I'll revert the properties and make `intensity` and `fraction` proper attributes again.